river/neighbors/base_neighbors.py

```diff
@@ -1,233 +1,201 @@
+        point.
 
     Notes
     -----
-    It updates its stored samples by the FIFO method, which means
+    Updates are stored by the FIFO (first in first out) method, which means
     that when size limit is reached, old samples are dumped to give
-    place to new samples.
-
-    The internal buffer does not keep order of the stored samples,
-    when the size limit is reached, the older samples are overwritten
-    with new ones (circular buffer).
+    place to new samples. This is circular, meaning that older points are dumped
+    first. This also gives the implementation a temporal aspect, because older
+    samples are replaced with newer ones. However, if min_distance_keep is set
+    to be != 0, every sample given to learn may not be added to the window.
 
     """
 
-    def __init__(self, window_size: int = 1000):
+    def __init__(self, window_size, p, min_distance_keep=0.05, save_uid=False):
         self.window_size = window_size
-        self._n_features: int = -1
-        self._n_targets: int = -1
-        self._size: int = 0
-        self._next_insert: int = 0
-        self._oldest: int = 0
-        self._imask: np.ndarray
-        self._X: np.ndarray
-        self._y: typing.List
-        self._is_initialized: bool = False
-
-    def _configure(self):
-        # Binary instance mask to filter data in the buffer
-        self._imask = np.zeros(self.window_size, dtype=bool)
-        self._X = np.zeros((self.window_size, self._n_features))
-        self._y = [None for _ in range(self.window_size)]
-        self._is_initialized = True
-
-    def reset(self):
-        """Reset the sliding window."""
-        self._n_features = -1
-        self._n_targets = -1
-        self._size = 0
-        self._next_insert = 0
-        self._oldest = 0
-        self._imask = None
-        self._X = None
-        self._y = None
-        self._is_initialized = False
-
-        return self
-
-    def append(self, x: np.ndarray, y: base.typing.Target) -> "KNeighborsBuffer":
-        """Add a (single) sample to the sample window.
-
-        x
-            1D-array of feature for a single sample.
-
-        y
-            The target data for a single sample.
+        self.p = p
+        self.window = collections.deque(maxlen=window_size)
 
-        Raises
-        ------
-        ValueError: If at any moment, a sample with a different number of
-        attributes than that of the n_attributes parameter is passed, a
-        ValueError is raised.
+        # A minimum distance (similarity) to determine adding to window
+        # The model will perform better with a more diverse window
+        self.min_distance_keep = min_distance_keep
 
-        TypeError: If the buffer type is altered by the user, or is not
-        correctly initialized, a TypeError may be raised.
+        # Save a unique identifier for each neighbor to return later
+        self.save_uid = save_uid
 
+    def add_to_window(self, x, y, uid=None):
```

Comment:
I would call this `append`.

Reply:
+1 to that.

This is what I meant by sending the example with the kDTree from sklearn, maybe I did not express myself effectively (English is not my native language hahah). Although it only takes arrays (X), you can anything with that. It is meant to perform search queries in different flavors. 


I think that `NearestNeighbors` could indeed become part of a collection of sorts, along with (Ordered) Sliding Windows, some types of (biased) reservoir buffers, and stuff like that. Just thinking out loud, we could even have something like `NearestNeighbors` which is agnostic to the data ingestion strategy. For instance:

- FIFO
- Reservoir sampling
- Ordered Values
- Set-esque addition

As long as the underlying data buffers have a `__getitem__` we are good to go.

Reply:
> Does that make sense?

Yes totally! 

And I love your ideas @smastelini - it's not your English being bad, I'm just very dense, haha - I didn't think through the idea that the base class could be general (but the actual classes that the user interacts with could still be consistent with what River generally provides). 

I'll get in some new updates for this PR tomorrow / later today - after I go back to sleep and wake up again :P